trade-module/src/main/java/io/edpn/backend/trade/adapter/persistence/repository/MybatisLocateSingleHopTradeRouteRepository.java

```diff
@@ -0,0 +1,118 @@
+                                  AND timestamp >= (now() - INTERVAL '1 hour' * #{maxPriceAgeHours})
+                                  AND buy_price IS NOT null),
+            
+                 sell_market AS (SELECT commodity_id,
+                                        station_id                                                             as sell_station_id,
+                                        arrival_distance                                                       AS sell_arrival_distance,
+                                        MAX(sell_price) OVER (PARTITION BY commodity_id)                       AS max_sell_price,
+                                        sell_price,
+                                        coordinates_geom,
+                                        arrival_distance,
+                                        demand,
+                                        ROW_NUMBER() OVER (PARTITION BY commodity_id ORDER BY sell_price DESC) AS rn
+                                 FROM latest_market_datum sell
+                                          INNER JOIN stations_in_range ON station_id = stations_in_range.id
+                                          INNER JOIN system on stations_in_range.system_id = system.id
+                                          INNER JOIN commodity_list ON commodity_id = commodity_list.id AND display_name IS NOT null
+                                 WHERE (demand = 0 OR demand > #{minDemand})
+                                   AND commodity_list.is_rare = false
+                                   AND timestamp >= (now() - INTERVAL '1 hour' * #{maxPriceAgeHours}))
+            
+            SELECT buy_market.commodity_id,
+                   buy_market.buy_station_id,
+                   buy_market.buy_price,
+                   buy_market.stock,
+                   sell_market.sell_station_id,
+                   sell_market.sell_price,
+                   sell_market.demand,
+                   (sell_market.sell_price - buy_market.buy_price)                                                          AS profit,
+                   st_3ddistance(buy_market.coordinates_geom, sell_market.coordinates_geom)                                 AS route_distance,
+                   st_3ddistance(st_makepoint(#{xCoordinate}, #{yCoordinate}, #{zCoordinate}), buy_market.coordinates_geom) AS distance_from_reference
+            
+            FROM buy_market
+                     INNER JOIN sell_market ON buy_market.commodity_id = sell_market.commodity_id
+                AND st_3ddwithin(sell_market.coordinates_geom, buy_market.coordinates_geom, #{maxRouteDistance})
+                AND 100 >= buy_market.rn AND 100 >= sell_market.rn
+                AND buy_market.buy_station_id != sell_market.sell_station_id
+            WHERE (sell_market.sell_price - buy_market.buy_price) > 5000
+            ORDER BY profit DESC, buy_market.arrival_distance + sell_arrival_distance
+            LIMIT 50
+            </script>""")
```

Comment:
Just as a side note: I _really_ don't think annotations and complicated queries are a natural or good match.

Reply:
Agree to disagree. We've already had this conversation about 15 times.
Keep query by function that uses it.